polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+        if rhs == var_id {
+            constrain_var(result_set, types, vars, lhs, var_type);
+            let in_result_set = result_set.requests.remove(lhs).unwrap();
+            assert_eq!(result_set.resolve_order.pop(), Some(lhs.to_string()));
+            request.constraints.extend(in_result_set.constraints);
+        }
+    }
+
+    result_set.requests.insert(var_id.to_string(), request);
+    result_set.resolve_order.push(var_id.to_string());
+}
+
+pub fn opt_pass(filter_plan: &mut FilterPlan, explain: bool) -> bool {
+    let mut optimized = false;
+
+    // Remove duplicate result set in a union.
+    let mut drop_plan = None;
+    'plans: for (i, result_set_a) in filter_plan.result_sets.iter().enumerate() {
+        for (j, result_set_b) in filter_plan.result_sets.iter().enumerate() {
+            if i != j && result_set_a == result_set_b {
+                drop_plan = Some(j);
+                break 'plans;
+            }
+        }
+    }
+    if let Some(plan_id) = drop_plan {
+        if explain {
+            eprintln!("* Removed duplicate result set.")
+        }
+        filter_plan.result_sets.remove(plan_id);
+        optimized = true;
+    }
+
+    // Possible future optimization ideas.
+    // * If two result sets are almost the same except for a single fetch
+    //   that only has a single field check and the field is different, we
+    //   can merge the two result sets and turn the field check into an `in`.
+    //   This is basically "un-expanding" either an `in` or and `or` from the policy.
+    //   This could be hard to find.
+    optimized
+}
+
+pub fn optimize(mut filter_plan: FilterPlan, explain: bool) -> FilterPlan {
+    if explain {
+        eprintln!("\nOptimizing...")
+    }
+    while opt_pass(&mut filter_plan, explain) {}
+    if explain {
+        eprintln!("Done\n")
+    }
+    filter_plan
+}
+
+pub fn build_filter_plan(
+    types: Types,
+    partial_results: PartialResults,
+    variable: &str,
+    class_tag: &str,
+) -> PolarResult<FilterPlan> {
+    let mut filter_plan = FilterPlan {
+        result_sets: vec![],
+    };
+    // @NOTE(steve): Just reading an env var here sucks (see all the stuff we had to do
+    // to get POLAR_LOG to work in all libs, wasm etc...) but that's what I'm doing today.
+    // At some point surface this info better.
+    let explain = std::env::var("POLAR_EXPLAIN").is_ok();
+
+    if explain {
+        eprintln!("\n===Data Filtering Query===");
+        eprintln!("\n==Bindings==")
+    }
+
+    for (i, result) in partial_results.iter().enumerate() {
+        let term = result.bindings.get(&Symbol::new(variable)).unwrap();
+        let exp = term.value().as_expression()?;
+        assert_eq!(exp.operator, Operator::And);
+
+        if explain {
+            eprintln!("  {}: {}", i, term.to_polar());
+        }
+
+        let var_info = process_result(exp);
+        let vars = collapse_vars(var_info);
+
+        if explain {
```

Comment:
Breaking this explain into a separate function would make the main control flow easier to follow.